Earthfile

```diff
@@ -0,0 +1,216 @@
+        ACCEPT_EULA=Y apt-get install msodbcsql17 mssql-tools && \
+        odbcinst -i -d -f /opt/microsoft/msodbcsql17/etc/odbcinst.ini
+
+    # not used for the moment as Exasol is not tested here
+    # RUN cd /opt && \
+    #     wget -q https://www.exasol.com/support/secure/attachment/111075/EXASOL_ODBC-6.2.9.tar.gz && \
+    #     tar xzf EXASOL_ODBC-6.2.9.tar.gz && \
+    #     echo "\n[EXASOL]\nDriver=`ls /opt/EXA*/lib/linux/x86_64/libexaodbc-uo2214lv1.so`\nThreading=2\n" >> /etc/odbcinst.ini && \
+    #     rm /opt/*.tar.gz
+
+python:
+    ARG PYTHON_VERSION=3.8.6
+    ARG ARROW_VERSION_RULE="<2.0.0"
+    ARG NUMPY_VERSION_RULE=""
+    ARG CONDA_EXTRA=""
+    FROM +os-base
+
+    RUN mamba create -y -q -n turbodbc-dev \
+        c-compiler \
+        make \
+        ninja \
+        cmake \
+        coveralls \
+        gmock \
+        gtest \
+        cxx-compiler \
+        mock \
+        pytest \
+        pytest-cov \
+        python=${PYTHON_VERSION} \
+        unixodbc \
+        boost-cpp \
+        numpy$NUMPY_VERSION_RULE \
+        pyarrow$ARROW_VERSION_RULE \
+        pybind11 \
+        $CONDA_EXTRA
+
+    RUN echo "conda activate turbodbc-dev" >> ~/.bashrc
+    RUN echo 'export UNIXODBC_INCLUDE_DIR=$CONDA_PREFIX/include' >> ~/.bashrc
+
+    RUN cd /opt && \
+        wget -q https://github.com/pybind/pybind11/archive/v2.6.2.tar.gz && \
+        tar xvf v2.6.2.tar.gz
+
+build:
+    ARG PYTHON_VERSION=3.8.6
+    ARG ARROW_VERSION_RULE="<2.0.0"
+    ARG NUMPY_VERSION_RULE=""
+    ARG CONDA_EXTRA=""
+    FROM --build-arg PYTHON_VERSION="$PYTHON_VERSION" \
+        --build-arg ARROW_VERSION_RULE="$ARROW_VERSION_RULE" \
+        --build-arg NUMPY_VERSION_RULE="$NUMPY_VERSION_RULE" \
+        --build-arg CONDA_EXTRA="$CONDA_EXTRA" \
+        +python
+
+    COPY . /src
+    # remove a potential available host build
+    RUN rm -rf /src/build && mkdir /src/build
+    WORKDIR /src/build
+
+    ENV ODBCSYSINI=/src/earthly/odbc
+    ENV TURBODBC_TEST_CONFIGURATION_FILES="query_fixtures_postgresql.json,query_fixtures_mssql.json,query_fixtures_mysql.json"
+    RUN ln -s /opt/pybind11-2.6.2 /src/pybind11
+
+    RUN bash -ic " \
+        cmake -DBOOST_ROOT=$CONDA_PREFIX -DBUILD_COVERAGE=ON \
+            -DCMAKE_INSTALL_PREFIX=./dist  \
+            -DPYTHON_EXECUTABLE=`which python` \
+            -GNinja .. && \
+        ninja && \
+        cmake --build . --target install && \
+        cd dist && \
+        python setup.py sdist \
+        "
+    SAVE ARTIFACT /src/build/dist/dist /dist
+
+test:
+    ARG PYTHON_VERSION=3.8.6
+    ARG ARROW_VERSION_RULE=""
+    ARG NUMPY_VERSION_RULE=""
+    ARG CONDA_EXTRA=""
+    FROM --build-arg PYTHON_VERSION="$PYTHON_VERSION" \
+        --build-arg ARROW_VERSION_RULE="$ARROW_VERSION_RULE" \
+        --build-arg NUMPY_VERSION_RULE="$NUMPY_VERSION_RULE" \
+        --build-arg CONDA_EXTRA="$CONDA_EXTRA" \
+        +build
+
+    WITH DOCKER --compose ../earthly/docker-compose.yml
```

Comment:
As this is the problematic line regarding the missing `docker-compose` in the image, can someone explain to me why this needs to run in the tested image and doesn't run outside of the image as the rest of earthly seems to do?

Reply:
I think they decided to do this inside the container as well as you have full control over the setup there. If outside you would be dependent on docker-compose to be available externally for example. Here they install docker-compose when needed.
This of course leads to the caching issue.

One thing I had during development was a `WITH DOCKER` step where the images where pulled before doing a `COPY` with the source. So the pulled images were part of the cached steps. This is only a workaround.

Maybe we can improve it with https://docs.earthly.dev/guides/shared-cache or with https://docs.earthly.dev/guides/advanced-local-caching. Would need more time to check this. As I am a bit short of time today and tomorrow I would suggest to open an issue to improve this afterwards.

Reply:
I think this is not an issue we can solve here now and also might have to see what the earthly devs do. So I think this here is not a merge blocker?

Reply:
imo, it's an improvement over the existing setup. therefore, I wouldn't consider it a blocker